File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/adapters/ProductFragmentPagerAdapter.java
package openfoodfacts.github.scrachx.openfood.views.adapters;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.viewpager2.adapter.FragmentStateAdapter;

import java.util.ArrayList;
import java.util.List;

import openfoodfacts.github.scrachx.openfood.fragments.BaseFragment;
import openfoodfacts.github.scrachx.openfood.models.State;

public class ProductFragmentPagerAdapter extends FragmentStateAdapter {
    private final List<BaseFragment> fragments;
    private final List<String> tabsTitles;

    public ProductFragmentPagerAdapter(FragmentActivity fragmentActivity) {
        super(fragmentActivity);
        this.fragments = new ArrayList<>();
        this.tabsTitles = new ArrayList<>();
    }

    public void addFragment(BaseFragment fragment, String tabTitle) {
        this.fragments.add(fragment);
        this.tabsTitles.add(tabTitle);
    }

    @NonNull
    @Override
    public Fragment createFragment(int i) {
        return fragments.get(i);
    }

    @Override
    public int getItemCount() {
        return fragments.size();
    }

    public CharSequence getPageTitle(int position) {
        return tabsTitles.get(position);
    }

    public void refresh(State state) {
        for (BaseFragment f : fragments) {
            if (f.isAdded()) {
                f.refreshView(state);
            }
        }
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/ProductFragmentPagerAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.product

import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import openfoodfacts.github.scrachx.openfood.features.shared.BaseFragment
import openfoodfacts.github.scrachx.openfood.models.ProductState

class ProductFragmentPagerAdapter(fragmentActivity: FragmentActivity) : FragmentStateAdapter(fragmentActivity) {
    private val fragments = mutableListOf<BaseFragment>()
    private val tabsTitles = mutableListOf<String>()

    operator fun plusAssign(entry: Pair<BaseFragment, String>) {
        fragments += entry.first
        tabsTitles += entry.second
    }

    fun add(entry: Pair<BaseFragment, String>) = plusAssign(entry)

    override fun createFragment(i: Int) = fragments[i]

    override fun getItemCount() = fragments.size

    fun getPageTitle(position: Int) = tabsTitles[position]

    fun refresh(productState: ProductState) =
            fragments.filter { it.isAdded }.forEach { it.refreshView(productState) }
}